homeassistant/components/climate/__init__.py

```diff
@@ -558,17 +564,75 @@ async def async_turn_on(self) -> None:
             if mode not in self.hvac_modes:
                 continue
             await self.async_set_hvac_mode(mode)
-            break
+            return
+
+        raise NotImplementedError
+
+    @property
+    def _is_turn_on_supported(self) -> bool:
+        if (
+            type(self).async_turn_on is not ClimateEntity.async_turn_on
+            or type(self).turn_on is not ClimateEntity.turn_on
+        ):
+            return True
+
+        if len(self.hvac_modes) == 2 and HVACMode.OFF in self.hvac_modes:
+            return True
+
+        for mode in (HVACMode.HEAT_COOL, HVACMode.HEAT, HVACMode.COOL):
+            if mode in self.hvac_modes:
+                return True
+
+        return False
+
+    def turn_off(self) -> None:
+        """Turn the entity off."""
+        raise NotImplementedError
 
     async def async_turn_off(self) -> None:
         """Turn the entity off."""
-        if hasattr(self, "turn_off"):
+        if type(self).turn_off is not ClimateEntity.turn_off:
             await self.hass.async_add_executor_job(self.turn_off)
             return
 
         # Fake turn off
         if HVACMode.OFF in self.hvac_modes:
             await self.async_set_hvac_mode(HVACMode.OFF)
+            return
+
+        raise NotImplementedError
+
+    @property
+    def _is_turn_off_supported(self) -> bool:
+        if (
+            type(self).async_turn_off is not ClimateEntity.async_turn_off
+            or type(self).turn_off is not ClimateEntity.turn_off
+        ):
+            return True
+
+        if HVACMode.OFF in self.hvac_modes:
+            return True
+
+        return False
+
+    def toggle(self) -> None:
+        """Toggle the entity."""
+        raise NotImplementedError
+
+    async def async_toggle(self) -> None:
+        """Toggle the entity."""
+        if type(self).toggle is not ClimateEntity.toggle:
+            await self.hass.async_add_executor_job(self.toggle)
+            return
```

Comment:
Why do we need to allow overriding these, can't we make them `@final`?

Reply:
Integrations interact with devices that often have some form of native TOGGLE command, it then might make sense to use it to the integration author, instead of relying on checking state and then sending TURN_ON or TURN_OFF.

As for the achieved functionality, you are right that it probably should never be actually necessary to override it. But the exact same argument applies to `async_turn_off`. There is an obvious use case for overriding `async_turn_on`, namely if your device provides a way to turn on without explicitly setting mode of operation (usually setting the one it had before being turned off). But there is no obvious necessity to ever override `async_turn_off` - its operation should be the same as setting mode to `HVACMode.OFF`, which is exactly what our base implementation does already.

The only reason I see someone might need to override `async_turn_off` is when they are dealing with some very weird device, that does not technically support turning off, but kind of does. For instance a device that only supports two modes: `[HVACMode.COOL, HVACMode.FAN_ONLY]`. It might make sense then to interpret the second one as a sort of pseudo-off. And then you would need to override `async_turn_off` and `async_toggle` as well. In general overriding `async_toggle` seems to be actually necessary in exactly the same (extremely rare) situations where overriding `async_turn_off` is necessary. So if we were to make those functions final, we should do the exact same to TURN_OFF.

And I feel like both are unnecessarily restrictive to integrations authors, while providing negligible benefits in general. Like I said the main reason to allow overriding is to more closely mirror in code what actual devices are typically supporting, and toggle seems like a common enough functionality.

Reply:
So to sum up, I think our advice for a typical climate integration should be:

- Override (async or sync) `turn_on` if your device supports native TURN_ON. Make sure that `hvac_mode` will (eventually) be correct as well.

- Override (async or sync) `turn_off` if your device supports native TURN_OFF. Make sure to set `hvac_mode` to `HVACMode.Off`. In `set_hvac_mode` make sure setting it to `HVACMode.Off` does the exact same thing (probably you can just call `turn_off` from there).

- Override (async or sync) `toggle` if your device supports native TOGGLE. In that case make sure that TURN_ON and TURN_OFF are supported as well.

Reply:
OK, this makes sense.
Can you please open and link to a PR of https://developers.home-assistant.io/docs/core/entity/climate which explains this?